src/ser.rs

```diff
@@ -16,86 +18,157 @@ pub trait Writer {
+#[cfg(feature = "heapless-v0-7")]
+impl<const N: usize> Writer for heapless_v0_7::Vec<u8, N> {
     type Error = Error;
-
     fn write_all(&mut self, buf: &[u8]) -> Result<()> {
-        let l = buf.len();
-        if self.slice.len() - self.index < l {
-            // This buffer will not fit in our slice
-            return Err(Error::SerializeBufferFull(self.index));
-        }
-        self.slice[self.index..][..l].copy_from_slice(buf);
-        self.index += l;
-        Ok(())
+        self.extend_from_slice(buf)
+            .or(Err(Error::SerializeBufferFull(self.len())))
     }
 }
 
-impl<'a, const N: usize> Writer for &'a mut crate::Bytes<N> {
+#[cfg(feature = "heapless-v0-8")]
+impl<const N: usize> Writer for heapless_v0_8::Vec<u8, N> {
     type Error = Error;
-
     fn write_all(&mut self, buf: &[u8]) -> Result<()> {
         self.extend_from_slice(buf)
-            .map_err(|_| Error::SerializeBufferFull(buf.len()))
+            .or(Err(Error::SerializeBufferFull(self.len())))
     }
 }
 
-pub struct Serializer<W>
-// where
-//     W: Writer,
-{
-    pub writer: W,
+impl<'a, T: Writer> Writer for &'a mut T {
+    type Error = T::Error;
+    fn write_all(&mut self, buf: &[u8]) -> Result<(), Self::Error> {
+        (**self).write_all(buf)
+    }
+}
+
+// #[derive(Debug)]
+// pub struct SliceWriter<'a> {
+//     slice: &'a mut [u8],
+//     index: usize,
+// }
+
+// impl<'a> SliceWriter<'a> {
+//     /// Wraps a mutable slice so it can be used as a `Writer`.
+//     pub fn new(slice: &'a mut [u8]) -> SliceWriter<'a> {
+//         SliceWriter { slice, index: 0 }
+//     }
+
+//     /// Returns the number of bytes written to the underlying slice.
+//     pub fn bytes_written(&self) -> usize {
+//         self.index
+//     }
+
+//     /// Returns the underlying slice.
+//     pub fn into_inner(self) -> &'a mut [u8] {
+//         self.slice
+//     }
+// }
+
+// impl<'a> Writer for SliceWriter<'a> {
+//     type Error = Error;
+
+//     fn write_all(&mut self, buf: &[u8]) -> Result<()> {
+//         let l = buf.len();
+//         if self.slice.len() - self.index < l {
+//             // This buffer will not fit in our slice
+//             return Err(Error::SerializeBufferFull(self.index));
+//         }
+//         self.slice[self.index..][..l].copy_from_slice(buf);
+//         self.index += l;
+//         Ok(())
+//     }
+// }
+
+// impl<'a, const N: usize> Writer for &'a mut crate::Bytes<N> {
+//     type Error = Error;
+
+//     fn write_all(&mut self, buf: &[u8]) -> Result<()> {
+//         self.extend_from_slice(buf)
+//             .map_err(|_| Error::SerializeBufferFull(buf.len()))
+//     }
+// }
+
+struct WrittenWriter<W> {
+    writer: W,
+    written: usize,
+}
+
+impl<W: Writer> Writer for WrittenWriter<W> {
+    type Error = W::Error;
+
+    fn write_all(&mut self, buf: &[u8]) -> core::result::Result<(), Self::Error> {
+        self.written += buf.len();
+        self.writer.write_all(buf)
```

Comment:
I think `written` should only be increased if `write_all` is successful.  I don’t think this is the case today (or maybe ever) but theoretically, serialization could continue after a failure.

Reply:
If we follow the `write_all` from the standard library we could also have a `write_all` failure that has written some bytes. This is also internal API, and can be adapted in the future if required.

If we want to think forward, we should probably have a only a `write` function, and have `write_all` be a default impl.

Reply:
While the API is internal, the behavior of `write_all` is part of the public API via `Serializer`.